### PR TITLE
Nicholette/scribing question client

### PR DESCRIPTION
**Pull request by nicholette**
React client for scribing question creation.

Rails controllers/ models /views have been previously merged in from #2297.

Requires #2396 to be merged first.

### CODE REVIEW

File: client/package.json [hunk 1]
```diff
@@ -71,6 +71,7 @@
     "intl": "^1.2.5",
     "jquery-ui": "^1.12.1",
     "json-loader": "^0.5.4",
+    "lodash": "^4.17.4",
```
Comment (jeremyyap) [inline client/package.json]: There are only a few uses of lodash in your PR. Is it possible to rewrite those to not use lodash?

File: client/app/lib/components/redux-form/FlatButton.js [hunk 2]
```diff
@@ -0,0 +1,7 @@
+import FlatButton from 'material-ui/FlatButton';
+import createComponent from './createComponent';
+import mapError from './mapError';
+
+const mapProps = props => ({ ...mapError(props) });
+
+export default createComponent(FlatButton, mapProps);
```
Comment (jeremyyap) [inline client/app/lib/components/redux-form/FlatButton.js]: Is this component used anywhere?

File: client/app/bundles/course/assessment/question/scribing/reducers/scribingQuestionReducer.jsx [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    maximum_grade: 0,
+    weight: 0,
+    skill_ids: [],
+    skills: [],
+    attachment_reference: {
+      name: '',
+      path: '',
+      updater_name: '',
+    },
+  },
```
Comment (jeremyyap) [inline client/app/bundles/course/assessment/question/scribing/reducers/scribingQuestionReducer.jsx]: I think not necessary to specify all the nested fields for initialState

File: client/app/bundles/course/assessment/question/scribing/reducers/scribingQuestionReducer.jsx [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        isLoading: false,
+        isSubmitting: false,
+        saveErrors: undefined,
+        error: false,
+      };
+    }
+    case actionTypes.CREATE_SCRIBING_QUESTION_SUCCESS:
+    case actionTypes.UPDATE_SCRIBING_QUESTION_SUCCESS: {
+      const { courseId, assessmentId } = action;
+      redirectToAssessment(courseId, assessmentId);
```
Comment (jeremyyap) [inline client/app/bundles/course/assessment/question/scribing/reducers/scribingQuestionReducer.jsx]: Can move this to actions, reducer should only be concerned with updating state

File: client/app/bundles/course/assessment/question/scribing/components/FileUploadField.jsx [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+          </div>)
+        }
+      />
+    );
+  }
+}
+
+FileUploadField.propTypes = propTypes;
+
+export default createComponent(FileUploadField, mapProps);
```
Comment (jeremyyap) [inline client/app/bundles/course/assessment/question/scribing/components/FileUploadField.jsx]: There is another redux-form file input component at `lib/components/redux-form/SingleFileInput.jsx`. That one is used for achievement image upload if i recall correctly. If they are similar enough, maybe can combine together?

File: client/app/bundles/course/assessment/question/scribing/constants.js [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+  'SUBMISSION_MESSAGE_CLEAR',
+  'SUBMIT_FORM_LOADING',
+  'SUBMIT_FORM_SUCCESS',
+  'SUBMIT_FORM_FAILURE',
+  'VALIDATION_ERRORS_SET',
+  'CREATE_RESPONSE_REQUEST',
+  'CREATE_RESPONSE_FAILURE',
+  'LOAD_RESPONSE_REQUEST',
+  'UPDATE_RESPONSE_REQUEST',
+  'UPDATE_RESPONSE_FAILURE',
```
Comment (jeremyyap) [inline client/app/bundles/course/assessment/question/scribing/constants.js]: Some of these action types seem to be unused.

File: client/app/bundles/course/assessment/question/scribing/reducers/scribingQuestionReducer.jsx (hunk 3 above)
Comment (nicholette) [inline client/app/bundles/course/assessment/question/scribing/reducers/scribingQuestionReducer.jsx]: hmm true

File: client/app/bundles/course/assessment/question/scribing/components/FileUploadField.jsx (hunk 5 above)
Comment (nicholette) [inline client/app/bundles/course/assessment/question/scribing/components/FileUploadField.jsx]: The input values of `lib/components/redux-form/SingleFileInput.jsx` is somewhat different from the FileUploadField. Will need more investigation/ tweaking before they can be combined.

File: client/app/bundles/course/assessment/question/scribing/components/LoadingIndicator.jsx [hunk 7]
```diff
@@ -0,0 +1,23 @@
+import React from 'react';
```
Comment (jeremyyap) [inline client/app/bundles/course/assessment/question/scribing/components/LoadingIndicator.jsx]: This file is identical to `lib/components/LoadingIndicator.jsx`?

File: client/app/bundles/course/assessment/question/scribing/responses.js [hunk 8]
```diff
@@ -0,0 +1 @@
+import '../scribing';
```
Comment (jeremyyap) [inline client/app/bundles/course/assessment/question/scribing/responses.js]: What is the purpose of this file?

Comment (nicholette) [inline client/app/bundles/course/assessment/question/scribing/responses.js]: Can be removed ><

File: config/locales/en/course/assessment/assessments.yml [hunk 9]
```diff
@@ -39,6 +39,7 @@ en:
             text_response: 'Text Response'
             file_upload: 'File Upload'
             programming: 'Programming'
+            scribing: 'Scribing'
```
Comment (jeremyyap) [inline config/locales/en/course/assessment/assessments.yml]: Unused key `course.assessment.assessments.show.new_question.scribing`

Reported by https://travis-ci.org/Coursemology/coursemology2/jobs/254783336

Comment (nicholette) [inline config/locales/en/course/assessment/assessments.yml]: That is for the create new question link in the assessment page which is not committed yet so we do not expose the scribing feature. Should I removed this first?

Comment (jeremyyap) [inline config/locales/en/course/assessment/assessments.yml]: This will cause Travis builds for future PRs to fail until the new question link is added, I think it's better to remove it first.

Comment (nicholette) [inline config/locales/en/course/assessment/assessments.yml]: Ok

File: client/app/api/course/assessment/question/scribing/Scribing.js [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+  * }
+   */
+
+  /**
+   * Fetches a Scribing question with initial attributes
+   *
+   * @return {Promise}
+   * success response: scribing_question
+   */
+  new() {
```
Comment (allenwq) [inline client/app/api/course/assessment/question/scribing/Scribing.js]: What attributes that we need to get from this `new` action ?  I think frontend should have all the necessary info already ?

Comment (nicholette) [inline client/app/api/course/assessment/question/scribing/Scribing.js]: Needed this to get the assessment skills to display as skills options. The other attributes are not important but I used the same json.jbuilder template to return the attributes.

Comment (allenwq) [inline client/app/api/course/assessment/question/scribing/Scribing.js]: I see. But we probably should use a skill end point instead, other question types will also need the skills.

Comment (nicholette) [inline client/app/api/course/assessment/question/scribing/Scribing.js]: Yea, agree ><

File: app/views/course/assessment/question/scribing/new.html.slim [hunk 11]
```diff
@@ -1,3 +1,4 @@
 - add_breadcrumb :new
+- template = 'course/assessment/question/scribing/scribing_question.json.jbuilder'
 = page_header
-div#scribing-question
+div#scribing-question data=(render partial: template)
```
Comment (jeremyyap) [inline app/views/course/assessment/question/scribing/new.html.slim]: Actually doesn't this already render the jbuilder partial that includes the skills? So you don't have to make another call to the server?

Comment (nicholette) [inline app/views/course/assessment/question/scribing/new.html.slim]: Oh no! It's remnants from my first implementation, forgot to remove it! Since editing the scribing question is already making an ajax call, might as well do the same thing for creating a new question.

File: client/app/api/course/assessment/question/scribing/Scribing.js (hunk 10 above)
Comment (nicholette) [inline client/app/api/course/assessment/question/scribing/Scribing.js]: @allenwq should this change be in another PR?

Comment (allenwq) [inline client/app/api/course/assessment/question/scribing/Scribing.js]: You can submit another PR and make this PR depends on it, we can merge that one first
